src/Symfony/Component/Serializer/DenormalizationResult.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer;
+
+final class DenormalizationResult
+{
+    private $denormalizedValue;
+    private $invariantViolations = [];
+
+    private function __construct()
+    {
+    }
+
+    public static function success($denormalizedValue): self
+    {
+        $result = new self();
+        $result->denormalizedValue = $denormalizedValue;
+
+        return $result;
+    }
+
+    /**
+     * @param array<string, array<InvariantViolation>> $invariantViolations
+     */
+    public static function failure(array $invariantViolations): self
+    {
+        $result = new self();
+        $result->invariantViolations = $invariantViolations;
```

Comment:
It could be interesting to validate that the list really contains only instances of `InvariantViolation`

Reply:
IMO this should be covered with static analysis tools such as PHPStan or Psalm, not with runtime assertions.

Reply:
Woops, indeed that's an epic fail :)

I still think this should not be a runtime assertion though (which would have not detected this mistake anyway). There is no static analysis on the Symfony codebase as far as I know, this means this requires me to write tests that covers those scenarios instead.

Reply:
> which would have not detected this mistake anyway

What do you mean ?
Asserts should be enable on development environment and only disable in production.
Meaning that when running your tests with a `php.ini` that enables the assertions an error would have been thrown.
In case the static method implementation was not clear:
```php
private static function areViolationsValid(array $violations): bool
{
    foreach ($violations as $violation) {
        if (!$violation instanceof InvariantViolation) {
            return false;
        }
    }

    return true;
}
```

Reply:
> > which would have not detected this mistake anyway
>
> What do you mean ?

I mean adding those assertions would have not prevented me from making this mistake because there are no tests to run that code as of now so I would have received no alerts about that anyway.

> Asserts should be enable on development environment and only disable in production.

I would not allow disabling runtime assertions in production: the point is to prevent the application from silently ignoring errors and running in an invalid state. IMO detecting those error scenarios is more efficiently done using static analysis if possible, with tests otherwise. Here we have no static analysis so I'll go with tests.